qp_klp/process_amplicon_job.py

```diff
@@ -0,0 +1,383 @@
+        else:
+            raise e
+
+    # perform sample-id validation against Qiita
+    projects = pipeline.get_project_info()
+
+    errors = []
+    sn_tid_map_by_project = {}
+
+    for project in projects:
+        # remove the qiita-id prepending the project_name
+        project_name = remove_qiita_id(project['project_name'])
+        qiita_id = project['qiita_id']
+
+        df = pipeline.mapping_file
+        df = df.loc[df['project_name'] == project_name]
+
+        # assume the BLANKS in the mapping-file are not prepended w/qiita-id
+        # or some other value. Confirmed w/wet-lab.
+        mf_samples = {s for s in df['sample_name'] if
+                      not s.startswith('BLANK')}
+
+        # collect needed info from Qiita here.
+        url = f'/api/v1/study/{qiita_id}/samples'
+        qsamples = qclient.get(url)
+        qsamples = {x.replace(f'{qiita_id}.', '') for x in qsamples}
+        tube_id_present = 'tube_id' in qclient.get(f'{url}/info')['categories']
+        if tube_id_present:
+            tids = qclient.get(f'{url}/categories=tube_id')['samples']
+
+        # compare the list of samples from the mapping file against the
+        # list of samples from Qiita.
+        sample_name_diff = mf_samples - qsamples
+        sn_tid_map_by_project[project_name] = None
+
+        if sample_name_diff:
+            # if tube_id is defined in the Qiita study, then any sample_names
+            # missing from the mapping-file may simply have a leading zero
+            # present.
+            if tube_id_present:
+                # strip any leading zeroes from the sample-ids. Note that
+                # if a sample-id has more than one leading zero, all of
+                # them will be removed.
+                mf_samples = {x.lstrip('0') for x in mf_samples}
+                # once any leading zeros have been removed, recalculate
+                # sample_name_diff before continuing processing.
+                sample_name_diff = mf_samples - qsamples
+
+        if sample_name_diff:
+            # before we report as an error, check tube_id
+            if tube_id_present:
+                # generate a map of sample_names to tube_ids for
+                # GenPrepFileJob.
+                sn_tid_map_by_project[project_name] = {
+                    y[0]: x.replace(f'{qiita_id}.', '') for x, y in
+                    tids.items()}
+                tids = set(sn_tid_map_by_project[project_name].keys())
+                tube_id_diff = mf_samples - tids
+                if not tube_id_diff:
+                    continue
+                len_tube_id_overlap = len(tube_id_diff)
+                tids_example = ', '.join(choices(list(tids), k=5))
+                error_tube_id = (
+                    f'tube_id in Qiita but {len_tube_id_overlap} missing '
+                    f'samples. Some samples from tube_id: {tids_example}.')
+            else:
+                error_tube_id = 'No tube_id column in Qiita.'
```

Comment:
I don't fully follow why we need these two blocks? Can they be merged into one?

Reply:
Good catch, ty!